main/acle.md

```diff
@@ -1072,6 +1074,32 @@ including the header:
 
 Including `<arm_sme.h>` also includes [`<arm_sve.h>`](#arm_sve.h).
 
+### Predefined Macros and Headers
+
+The macros can represent a compiler's ability to use the instructions
+associated with the feature via ACLE documented intrinsics and within inline
+assembly. However users should not rely solely on macro definition, because
+there are intrinsics that may be influenced by mechanisms like `#pragma` or
+`attribute`. The macros are defined only for user convenience.
+
+For example:
+
+``` c
+   #ifdef __ARM_FEATURE_SME
+   #include<arm_sme.h>
+     void foo(svbool_t pg, void *ptr, uint32_t slice_base) {
+        svst1_hor_za8(0, slice_base, pg, ptr);
+      }
+   #endif
+```
+
+`foo`  uses `__ARM_FEATURE_SME` to have available SME intriniscs,
+like `svst1_hor_za8`. However it has the compilation error:
+`builtin can only be called from a streaming function`
+for the command line:
+`armclang -O3 -target aarch64-linux-gnu -march=armv8-a+sme`
+that defines `__ARM_FEATURE_SME` to one.
+
```

Comment:
How about?

Testing a feature macro returns the compiler's support for the feature via ACLE documented intrinsics and inline assembly. This property can be independent of where the test is performed, meaning a feature macro does not guarantee the validity of using the feature at the point it is tested. For example:

``` c
    #include <arm_sme.h>

    void foo(svbool_t pg, void *ptr, uint32_t slice_base) {
    #ifdef __ARM_FEATURE_SME
      svst1_hor_za8(0, slice_base, pg, ptr);
    #endif
    }
```

Whilst testing \`__ARM_FEATURE_SME\` ensures the compiler supports the SME intrinsic \`svst1_hor_za8\`, \`foo\` will fail to compile because it has not been decorated with the necessary keywords to guarantee the function will be executed in streaming mode.

Not all such issues can be caught during compilation and may instead result in runtime failures.

Reply:
> Testing a feature macro returns the compiler's support for the feature via ACLE documented intrinsics and inline assembly.

Not sure if it is just me, the key is how we define "support".
For me compiler supports ACLE when the given ACLE is implemented. While the feature macro is only defined when appropriate features are enabled in the cmd line option only. Maybe worth to change it to sth like:

"Testing a feature macro allows to check if the compiler can make use of the architecture feature via ACLE documented intrinsics and inline assembly assuming the provided compilation flags."

What do you think?

Reply:
This distinction is not relevant because there is no assumption about compilation flags.  The macro alone is sufficient to detect at the source level whether the compiler (in the mode it's in when compiling the input) can support a specific feature.  By support we effectively mean "is able to compile the input when it is correctly formed".

The provided example shows something that is not "correctly formed" where the compiler will report an error that must be fixed before the code will compile. There are many other examples but I don't see a reason to document them all here.

Note I say "compile" and not "compile to a working binary" because the more typical result is the compiler correctly building the input which then generates a SIGILL when executed. This is covered by the "Not all such issues..." sentence.  Personally I don't think it's worth going into anymore detail than this.

Reply:
Sure, I get the difference between "compile" and not "compile to a working binary". 

For me it was just the meaning of word support, as in my mind feature could be supported (as implemented) by a tool but not available when given set of options it provided. It has this small difference from the user point of view as if feature is not implemented they can not do anything, while if it is not enabled they can change options.
But I am not against your version either.

Reply:
I think that's because you are considering this from the point of view of a human asking whether the compiler can support a feature. Whereas the feature macros are source code actions and thus the context is the point the source code is being compiled, at which point the compiler has already been "configured" (for want of a better word) by the user.

Perhaps the point you are raising is the document does not detail what actions the user should take in order to control the definitions of the feature flags, which is intentional because we cannot dictate that.

Reply:
I agree it would be good to be a bit more precise about what “support” means though.  At one level, GCC 14 “supports” SME2, and the word is often used in that sense.  So if we just say “support”, users might expect GCC 14 to define `__ARM_FEATURE_SME2` unconditionally.

Instead, for compatibility, we have to define the macro only if the intrinsics are “available” in functions that have no target annotations (not precise wording either).  What users need to do to achieve that for a given compiler is outside the scope of the ACLE, but I think the result is within scope.

IMO, the example of using SME intrinsics in non-streaming functions is an example of an intrinsics being available, but being used incorrectly.  It's somewhat like passing an out-of-range value to a constant argument, except that the error is contextual, rather than contained within the call.  (I still think it's a good example though.)

Reply:
I did changed to what @paulwalker-arm requested, but instead of using support I wrote available, as suggested by @rsandifo-arm.
I hope it is correct.